
.. DO NOT EDIT.
.. THIS FILE WAS AUTOMATICALLY GENERATED BY SPHINX-GALLERY.
.. TO MAKE CHANGES, EDIT THE SOURCE PYTHON FILE:
.. "examples/05_model_perturbation_hook.py"
.. LINE NUMBERS ARE GIVEN BELOW.

.. only:: html

    .. note::
        :class: sphx-glr-download-link-note

        :ref:`Go to the end <sphx_glr_download_examples_05_model_perturbation_hook.py>`
        to download the full example code

.. rst-class:: sphx-glr-example-title

.. _sphx_glr_examples_05_model_perturbation_hook.py:


Model Hook Injection: Perturbation
==========================

Adding model noise by using custom hooks.

This example will demonstrate how to run a an ensemble inference workflow to generate a
perturbed ensemble forecast. This perturbation is done by injecting code into the model
front and rear hooks. These hooks are applied to the tensor data before/after the model forward call.

This example also illustrates how you can subselect data for IO. In this example we will only output
two variables: total column water vapour (tcwv) and 500 hPa geopotential (z500). To run this make
sure that the model selected predicts these variables are change appropriately.

In this example you will learn:

- How to instantiate a built in prognostic model
- Creating a data source and IO object
- Changing the model forward/rear hooks
- Choose a subselection of coordinates to save to an IO object.
- Post-processing results

.. GENERATED FROM PYTHON SOURCE LINES 42-61

Creating an Ensemble Workflow
-----------------------------------

To start lets begin with creating an ensemble workflow to use. We encourage
users to explore and experiment with their own custom workflows that borrow ideas from
built in workflows inside :py:obj:`earth2studio.run` or the examples.

Creating our own generalizable ensemble workflow is easy when we rely on the component
interfaces defined in Earth2Studio (use dependency injection). Here we create a run
method that accepts the following:

- time: Input list of datetimes / strings to run inference for
- nsteps: Number of forecast steps to predict
- nensemble: Number of ensembles to run for
- prognostic: Our initialized prognostic model
- data: Initialized data source to fetch initial conditions from
- io: io store that data is written to.
- output_coords: CoordSystem of output coordinates that should be saved. Should be
     a proper subset of model output coordinates.

.. GENERATED FROM PYTHON SOURCE LINES 63-168

.. code-block:: Python

    from collections import OrderedDict
    from datetime import datetime
    from dotenv import load_dotenv

    load_dotenv()  # TODO: make common example prep function

    import numpy as np
    import torch
    from loguru import logger
    from tqdm import tqdm

    from earth2studio.data import DataSource, fetch_data
    from earth2studio.io import IOBackend
    from earth2studio.models.px import PrognosticModel
    from earth2studio.utils.coords import CoordSystem, map_coords, extract_coords
    from earth2studio.utils.time import to_time_array

    logger.remove()
    logger.add(lambda msg: tqdm.write(msg, end=""), colorize=True)


    def run_ensemble(
        time: list[str] | list[datetime] | list[np.datetime64],
        nsteps: int,
        nensemble: int,
        prognostic: PrognosticModel,
        data: DataSource,
        io: IOBackend,
        output_coords: CoordSystem = OrderedDict({}),
    ) -> IOBackend:
        """Ensemble workflow

        Parameters
        ----------
        time : list[str] | list[datetime] | list[np.datetime64]
            List of string, datetimes or np.datetime64
        nsteps : int
            Number of forecast steps
        nensemble : int
            Number of ensemble members to run inference for.
        prognostic : PrognosticModel
            Prognostic models
        data : DataSource
            Data source
        io : IOBackend
            IO object

        Returns
        -------
        IOBackend
            Output IO object
        """
        logger.info("Running ensemble inference!")

        # Load model onto the device
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        logger.info(f"Inference device: {device}")
        prognostic = prognostic.to(device)

        # Fetch data from data source and load onto device
        time = to_time_array(time)
        x, coords = fetch_data(
            source=data,
            time=time,
            lead_time=prognostic.input_coords["lead_time"],
            variable=prognostic.input_coords["variable"],
            device=device,
        )
        logger.success(f"Fetched data from {data.__class__.__name__}")

        # Expand x, coords for ensemble
        x = x.unsqueeze(0).repeat(nensemble, *([1] * x.ndim))
        coords = {"ensemble": np.arange(nensemble)} | coords

        # Set up IO backend with information from output_coords (if applicable).
        total_coords = coords.copy()
        total_coords["lead_time"] = np.asarray(
            [prognostic.output_coords["lead_time"] * i for i in range(nsteps + 1)]
        ).flatten()
        for key, value in total_coords.items():
            total_coords[key] = output_coords.get(key, value)

        variables_to_save = total_coords.pop("variable")
        io.add_array(total_coords, variables_to_save)

        # Map lat and lon if needed
        x, coords = map_coords(x, coords, prognostic.input_coords)

        # Create prognostic iterator
        model = prognostic.create_iterator(x, coords)

        logger.info("Inference starting!")
        with tqdm(total=nsteps + 1, desc="Running inference") as pbar:
            for step, (x, coords) in enumerate(model):
                # Subselect domain/variables as indicated in output_coords
                x, coords = map_coords(x, coords, output_coords)
                io.write(*extract_coords(x, coords))
                pbar.update(1)
                if step == nsteps:
                    break

        logger.success("Inference complete")
        return io









.. GENERATED FROM PYTHON SOURCE LINES 169-183

Set Up
------
With the ensemble workflow defined, we now need to create the indivdual components.

We need the following:

- Prognostic Model: Use the built in DLWP model :py:class:`earth2studio.models.px.DLWP`.
- Datasource: Pull data from the GFS data api :py:class:`earth2studio.data.GFS`.
- IO Backend: Lets save the outputs into a Zarr store :py:class:`earth2studio.io.ZarrBackend`.

We will first run the ensemble workflow using an unmodified function, that is a model that has the
default (identity) forward and rear hooks. Then we will define new hooks for the model and rerun the
inference request.
%%

.. GENERATED FROM PYTHON SOURCE LINES 183-202

.. code-block:: Python

    import numpy as np
    import torch

    from earth2studio.models.px import DLWP
    from earth2studio.data import GFS
    from earth2studio.io import ZarrBackend

    # Load the default model package which downloads the check point from NGC
    package = DLWP.load_default_package()
    model = DLWP.load_model(package)

    # Create the data source
    data = GFS()

    # Create the IO handler, store in memory
    chunks = {"ensemble": 1, "time": 1}
    io_unperturbed = ZarrBackend(file_name="outputs/ensemble.zarr", chunks=chunks)









.. GENERATED FROM PYTHON SOURCE LINES 203-211

Execute the Workflow
--------------------
With all componments intialized, running the workflow is a single line of Python code.
Workflow will return the provided IO object back to the user, which can be used to
then post process. Some have additional APIs that can be handy for post-processing or
saving to file. Check the API docs for more information.

%%

.. GENERATED FROM PYTHON SOURCE LINES 211-258

.. code-block:: Python


    nsteps = 4 * 12
    nensemble = 16
    forecast_date = "2024-01-30"
    output_coords = {
        "lat": np.arange(25.0, 60.0, 0.25),
        "lon": np.arange(230.0, 300.0, 0.25),
        "variable": np.array(["tcwv", "z500"]),
    }

    # Forst run the unperturbed model forcast
    io_unperturbed = run_ensemble(
        [forecast_date],
        nsteps,
        nensemble,
        model,
        data,
        io_unperturbed,
        output_coords=output_coords,
    )

    # Introduce slight model perturbation
    # front_hook / rear_hook map (x, coords) -> (x, coords)
    model.front_hook = lambda x, coords: (
        x
        - 0.05
        * x.var(dim=0)
        * (x - model.center.unsqueeze(-1))
        / (model.scale.unsqueeze(-1)) ** 2
        + 0.1 * (x - x.mean(dim=0)),
        coords,
    )
    # Also could use model.rear_hook = ...

    io_perturbed = ZarrBackend(
        file_name="outputs/ensemble_model_perturbation.zarr", chunks=chunks
    )
    io_perturbed = run_ensemble(
        [forecast_date],
        nsteps,
        nensemble,
        model,
        data,
        io_perturbed,
        output_coords=output_coords,
    )





.. rst-class:: sphx-glr-script-out

 .. code-block:: none

    2024-04-12 23:57:23.311 | INFO     | __main__:run_ensemble:115 - Running ensemble inference!
    2024-04-12 23:57:23.311 | INFO     | __main__:run_ensemble:119 - Inference device: cuda
    2024-04-12 23:57:23.318 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:151 - Fetching GFS index file: 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:23.658 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t850 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]    Fetching GFS for 2024-01-29 18:00:00:  14%|█▍        | 1/7 [00:00<00:04,  1.43it/s]                                                                                       2024-04-12 23:57:24.355 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z1000 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:  14%|█▍        | 1/7 [00:00<00:04,  1.43it/s]    Fetching GFS for 2024-01-29 18:00:00:  29%|██▊       | 2/7 [00:01<00:03,  1.63it/s]                                                                                       2024-04-12 23:57:24.910 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z700 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:  29%|██▊       | 2/7 [00:01<00:03,  1.63it/s]    Fetching GFS for 2024-01-29 18:00:00:  43%|████▎     | 3/7 [00:01<00:02,  1.88it/s]                                                                                       2024-04-12 23:57:25.346 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z500 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:  43%|████▎     | 3/7 [00:01<00:02,  1.88it/s]    Fetching GFS for 2024-01-29 18:00:00:  57%|█████▋    | 4/7 [00:02<00:01,  2.04it/s]                                                                                       2024-04-12 23:57:25.770 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z300 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:  57%|█████▋    | 4/7 [00:02<00:01,  2.04it/s]    Fetching GFS for 2024-01-29 18:00:00:  71%|███████▏  | 5/7 [00:02<00:00,  2.28it/s]                                                                                       2024-04-12 23:57:26.116 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: tcwv at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:  71%|███████▏  | 5/7 [00:02<00:00,  2.28it/s]    Fetching GFS for 2024-01-29 18:00:00:  86%|████████▌ | 6/7 [00:03<00:00,  2.09it/s]                                                                                       2024-04-12 23:57:26.672 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t2m at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:  86%|████████▌ | 6/7 [00:03<00:00,  2.09it/s]    Fetching GFS for 2024-01-29 18:00:00: 100%|██████████| 7/7 [00:03<00:00,  1.91it/s]    Fetching GFS for 2024-01-29 18:00:00: 100%|██████████| 7/7 [00:03<00:00,  1.93it/s]
    2024-04-12 23:57:27.293 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:151 - Fetching GFS index file: 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:27.387 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t850 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]    Fetching GFS for 2024-01-30 00:00:00:  14%|█▍        | 1/7 [00:00<00:02,  2.27it/s]                                                                                       2024-04-12 23:57:27.828 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z1000 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:  14%|█▍        | 1/7 [00:00<00:02,  2.27it/s]    Fetching GFS for 2024-01-30 00:00:00:  29%|██▊       | 2/7 [00:00<00:02,  2.25it/s]                                                                                       2024-04-12 23:57:28.274 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z700 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:  29%|██▊       | 2/7 [00:00<00:02,  2.25it/s]    Fetching GFS for 2024-01-30 00:00:00:  43%|████▎     | 3/7 [00:01<00:01,  2.33it/s]                                                                                       2024-04-12 23:57:28.685 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z500 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:  43%|████▎     | 3/7 [00:01<00:01,  2.33it/s]    Fetching GFS for 2024-01-30 00:00:00:  57%|█████▋    | 4/7 [00:01<00:01,  2.26it/s]                                                                                       2024-04-12 23:57:29.147 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z300 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:  57%|█████▋    | 4/7 [00:01<00:01,  2.26it/s]    Fetching GFS for 2024-01-30 00:00:00:  71%|███████▏  | 5/7 [00:02<00:00,  2.33it/s]                                                                                       2024-04-12 23:57:29.552 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: tcwv at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:  71%|███████▏  | 5/7 [00:02<00:00,  2.33it/s]    Fetching GFS for 2024-01-30 00:00:00:  86%|████████▌ | 6/7 [00:02<00:00,  2.17it/s]                                                                                       2024-04-12 23:57:30.073 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t2m at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:  86%|████████▌ | 6/7 [00:02<00:00,  2.17it/s]    Fetching GFS for 2024-01-30 00:00:00: 100%|██████████| 7/7 [00:03<00:00,  2.40it/s]    Fetching GFS for 2024-01-30 00:00:00: 100%|██████████| 7/7 [00:03<00:00,  2.32it/s]
    2024-04-12 23:57:30.438 | SUCCESS  | __main__:run_ensemble:131 - Fetched data from GFS
    2024-04-12 23:57:30.461 | INFO     | __main__:run_ensemble:154 - Inference starting!
    Running inference:   0%|          | 0/49 [00:00<?, ?it/s]    Running inference:   4%|▍         | 2/49 [00:00<00:06,  7.83it/s]    Running inference:   6%|▌         | 3/49 [00:00<00:05,  7.93it/s]    Running inference:   8%|▊         | 4/49 [00:00<00:06,  6.46it/s]    Running inference:  10%|█         | 5/49 [00:00<00:06,  7.05it/s]    Running inference:  12%|█▏        | 6/49 [00:00<00:06,  6.24it/s]    Running inference:  14%|█▍        | 7/49 [00:01<00:06,  6.41it/s]    Running inference:  16%|█▋        | 8/49 [00:01<00:07,  5.56it/s]    Running inference:  18%|█▊        | 9/49 [00:01<00:07,  5.70it/s]    Running inference:  20%|██        | 10/49 [00:01<00:07,  5.29it/s]    Running inference:  22%|██▏       | 11/49 [00:01<00:07,  5.39it/s]    Running inference:  24%|██▍       | 12/49 [00:02<00:07,  4.89it/s]    Running inference:  27%|██▋       | 13/49 [00:02<00:07,  5.13it/s]    Running inference:  29%|██▊       | 14/49 [00:02<00:07,  4.80it/s]    Running inference:  31%|███       | 15/49 [00:02<00:06,  5.04it/s]    Running inference:  33%|███▎      | 16/49 [00:02<00:07,  4.61it/s]    Running inference:  35%|███▍      | 17/49 [00:03<00:06,  4.74it/s]    Running inference:  37%|███▋      | 18/49 [00:03<00:07,  4.42it/s]    Running inference:  39%|███▉      | 19/49 [00:03<00:06,  4.54it/s]    Running inference:  41%|████      | 20/49 [00:03<00:06,  4.33it/s]    Running inference:  43%|████▎     | 21/49 [00:04<00:06,  4.58it/s]    Running inference:  45%|████▍     | 22/49 [00:04<00:06,  4.31it/s]    Running inference:  47%|████▋     | 23/49 [00:04<00:05,  4.48it/s]    Running inference:  49%|████▉     | 24/49 [00:04<00:05,  4.27it/s]    Running inference:  51%|█████     | 25/49 [00:04<00:05,  4.46it/s]    Running inference:  53%|█████▎    | 26/49 [00:05<00:05,  4.19it/s]    Running inference:  55%|█████▌    | 27/49 [00:05<00:04,  4.43it/s]    Running inference:  57%|█████▋    | 28/49 [00:05<00:05,  4.13it/s]    Running inference:  59%|█████▉    | 29/49 [00:05<00:04,  4.26it/s]    Running inference:  61%|██████    | 30/49 [00:06<00:04,  3.89it/s]    Running inference:  63%|██████▎   | 31/49 [00:06<00:04,  4.05it/s]    Running inference:  65%|██████▌   | 32/49 [00:06<00:04,  3.86it/s]    Running inference:  67%|██████▋   | 33/49 [00:06<00:03,  4.01it/s]    Running inference:  69%|██████▉   | 34/49 [00:07<00:03,  3.83it/s]    Running inference:  71%|███████▏  | 35/49 [00:07<00:03,  4.00it/s]    Running inference:  73%|███████▎  | 36/49 [00:07<00:03,  3.66it/s]    Running inference:  76%|███████▌  | 37/49 [00:08<00:03,  3.66it/s]    Running inference:  78%|███████▊  | 38/49 [00:08<00:03,  3.51it/s]    Running inference:  80%|███████▉  | 39/49 [00:08<00:02,  3.71it/s]    Running inference:  82%|████████▏ | 40/49 [00:08<00:02,  3.58it/s]    Running inference:  84%|████████▎ | 41/49 [00:09<00:02,  3.71it/s]    Running inference:  86%|████████▌ | 42/49 [00:09<00:01,  3.54it/s]    Running inference:  88%|████████▊ | 43/49 [00:09<00:01,  3.64it/s]    Running inference:  90%|████████▉ | 44/49 [00:10<00:01,  3.40it/s]    Running inference:  92%|█████████▏| 45/49 [00:10<00:01,  3.44it/s]    Running inference:  94%|█████████▍| 46/49 [00:10<00:00,  3.21it/s]    Running inference:  96%|█████████▌| 47/49 [00:11<00:00,  3.31it/s]    Running inference:  98%|█████████▊| 48/49 [00:11<00:00,  3.14it/s]    Running inference: 100%|██████████| 49/49 [00:11<00:00,  3.26it/s]    Running inference: 100%|██████████| 49/49 [00:11<00:00,  4.20it/s]
    2024-04-12 23:57:42.123 | SUCCESS  | __main__:run_ensemble:164 - Inference complete
    2024-04-12 23:57:42.123 | INFO     | __main__:run_ensemble:115 - Running ensemble inference!
    2024-04-12 23:57:42.124 | INFO     | __main__:run_ensemble:119 - Inference device: cuda
    2024-04-12 23:57:42.125 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:151 - Fetching GFS index file: 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.536 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t850 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.557 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z1000 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.575 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z700 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.594 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z500 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.613 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z300 at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.631 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: tcwv at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:   0%|          | 0/7 [00:00<?, ?it/s]    Fetching GFS for 2024-01-29 18:00:00:  86%|████████▌ | 6/7 [00:00<00:00, 53.06it/s]                                                                                       2024-04-12 23:57:42.649 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t2m at 2024-01-29 18:00:00
    Fetching GFS for 2024-01-29 18:00:00:  86%|████████▌ | 6/7 [00:00<00:00, 53.06it/s]    Fetching GFS for 2024-01-29 18:00:00: 100%|██████████| 7/7 [00:00<00:00, 53.29it/s]
    2024-04-12 23:57:42.674 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:151 - Fetching GFS index file: 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.753 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t850 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.780 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z1000 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.798 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z700 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.816 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z500 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.834 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: z300 at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]                                                                               2024-04-12 23:57:42.852 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: tcwv at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:   0%|          | 0/7 [00:00<?, ?it/s]    Fetching GFS for 2024-01-30 00:00:00:  86%|████████▌ | 6/7 [00:00<00:00, 51.68it/s]                                                                                       2024-04-12 23:57:42.870 | DEBUG    | earth2studio.data.gfs:fetch_gfs_dataarray:197 - Fetching GFS grib file for variable: t2m at 2024-01-30 00:00:00
    Fetching GFS for 2024-01-30 00:00:00:  86%|████████▌ | 6/7 [00:00<00:00, 51.68it/s]    Fetching GFS for 2024-01-30 00:00:00: 100%|██████████| 7/7 [00:00<00:00, 52.03it/s]
    2024-04-12 23:57:42.924 | SUCCESS  | __main__:run_ensemble:131 - Fetched data from GFS
    2024-04-12 23:57:42.944 | INFO     | __main__:run_ensemble:154 - Inference starting!
    Running inference:   0%|          | 0/49 [00:00<?, ?it/s]    Running inference:   4%|▍         | 2/49 [00:00<00:05,  8.91it/s]    Running inference:   6%|▌         | 3/49 [00:00<00:05,  8.86it/s]    Running inference:   8%|▊         | 4/49 [00:00<00:06,  7.07it/s]    Running inference:  10%|█         | 5/49 [00:00<00:06,  7.19it/s]    Running inference:  12%|█▏        | 6/49 [00:00<00:07,  6.13it/s]    Running inference:  14%|█▍        | 7/49 [00:01<00:06,  6.36it/s]    Running inference:  16%|█▋        | 8/49 [00:01<00:07,  5.71it/s]    Running inference:  18%|█▊        | 9/49 [00:01<00:06,  5.97it/s]    Running inference:  20%|██        | 10/49 [00:01<00:07,  5.33it/s]    Running inference:  22%|██▏       | 11/49 [00:01<00:06,  5.55it/s]    Running inference:  24%|██▍       | 12/49 [00:02<00:07,  5.06it/s]    Running inference:  27%|██▋       | 13/49 [00:02<00:06,  5.25it/s]    Running inference:  29%|██▊       | 14/49 [00:02<00:07,  4.85it/s]    Running inference:  31%|███       | 15/49 [00:02<00:06,  5.09it/s]    Running inference:  33%|███▎      | 16/49 [00:02<00:06,  4.72it/s]    Running inference:  35%|███▍      | 17/49 [00:03<00:06,  4.91it/s]    Running inference:  37%|███▋      | 18/49 [00:03<00:06,  4.58it/s]    Running inference:  39%|███▉      | 19/49 [00:03<00:06,  4.76it/s]    Running inference:  41%|████      | 20/49 [00:03<00:06,  4.47it/s]    Running inference:  43%|████▎     | 21/49 [00:03<00:06,  4.61it/s]    Running inference:  45%|████▍     | 22/49 [00:04<00:06,  4.35it/s]    Running inference:  47%|████▋     | 23/49 [00:04<00:05,  4.56it/s]    Running inference:  49%|████▉     | 24/49 [00:04<00:05,  4.25it/s]    Running inference:  51%|█████     | 25/49 [00:04<00:05,  4.41it/s]    Running inference:  53%|█████▎    | 26/49 [00:05<00:05,  4.15it/s]    Running inference:  55%|█████▌    | 27/49 [00:05<00:05,  4.35it/s]    Running inference:  57%|█████▋    | 28/49 [00:05<00:05,  4.09it/s]    Running inference:  59%|█████▉    | 29/49 [00:05<00:04,  4.24it/s]    Running inference:  61%|██████    | 30/49 [00:06<00:04,  3.92it/s]    Running inference:  63%|██████▎   | 31/49 [00:06<00:04,  4.03it/s]    Running inference:  65%|██████▌   | 32/49 [00:06<00:04,  3.75it/s]    Running inference:  67%|██████▋   | 33/49 [00:06<00:04,  3.82it/s]    Running inference:  69%|██████▉   | 34/49 [00:07<00:04,  3.60it/s]    Running inference:  71%|███████▏  | 35/49 [00:07<00:03,  3.69it/s]    Running inference:  73%|███████▎  | 36/49 [00:07<00:03,  3.49it/s]    Running inference:  76%|███████▌  | 37/49 [00:08<00:03,  3.59it/s]    Running inference:  78%|███████▊  | 38/49 [00:08<00:03,  3.40it/s]    Running inference:  80%|███████▉  | 39/49 [00:08<00:02,  3.46it/s]    Running inference:  82%|████████▏ | 40/49 [00:09<00:02,  3.27it/s]    Running inference:  84%|████████▎ | 41/49 [00:09<00:02,  3.37it/s]    Running inference:  86%|████████▌ | 42/49 [00:09<00:02,  3.22it/s]    Running inference:  88%|████████▊ | 43/49 [00:09<00:01,  3.34it/s]    Running inference:  90%|████████▉ | 44/49 [00:10<00:01,  3.18it/s]    Running inference:  92%|█████████▏| 45/49 [00:10<00:01,  3.27it/s]    Running inference:  94%|█████████▍| 46/49 [00:10<00:00,  3.12it/s]    Running inference:  96%|█████████▌| 47/49 [00:11<00:00,  3.18it/s]    Running inference:  98%|█████████▊| 48/49 [00:11<00:00,  3.05it/s]    Running inference: 100%|██████████| 49/49 [00:11<00:00,  3.17it/s]    Running inference: 100%|██████████| 49/49 [00:11<00:00,  4.13it/s]
    2024-04-12 23:57:54.815 | SUCCESS  | __main__:run_ensemble:164 - Inference complete




.. GENERATED FROM PYTHON SOURCE LINES 259-266

Post Processing
---------------
The last step is to post process our results. Cartopy is a greate library for plotting
fields on projects of a sphere. Here we plot and compare the ensemble mean and standard
deviation from using a unperturbed/perturbed model.

Notice that the Zarr IO function has additional APIs to interact with the stored data.

.. GENERATED FROM PYTHON SOURCE LINES 268-400

.. code-block:: Python

    import matplotlib.pyplot as plt
    import cartopy.crs as ccrs
    import matplotlib.animation as animation
    from matplotlib.colors import LogNorm


    levels_unperturbed = np.linspace(0, io_unperturbed["tcwv"][:].max())
    levels_perturbed = np.linspace(0, io_perturbed["tcwv"][:].max())

    std_levels_perturbed = np.linspace(0, io_perturbed["tcwv"][:].std(axis=0).max())

    plt.close("all")
    fig = plt.figure(figsize=(20, 10), tight_layout=True)
    ax0 = fig.add_subplot(2, 2, 1, projection=ccrs.PlateCarree())
    ax1 = fig.add_subplot(2, 2, 2, projection=ccrs.PlateCarree())
    ax2 = fig.add_subplot(2, 2, 3, projection=ccrs.PlateCarree())
    ax3 = fig.add_subplot(2, 2, 4, projection=ccrs.PlateCarree())


    def update(frame):
        """This function updates the frame with a new lead time for animation."""
        ax0.clear()
        ax1.clear()
        ax2.clear()
        ax3.clear()

        ## Update unperturbed image
        im0 = ax0.contourf(
            io_unperturbed["lon"][:],
            io_unperturbed["lat"][:],
            io_unperturbed["tcwv"][:, 0, frame].mean(axis=0),
            transform=ccrs.PlateCarree(),
            cmap="Blues",
            levels=levels_unperturbed,
        )
        ax0.coastlines()
        ax0.gridlines()

        im1 = ax1.contourf(
            io_unperturbed["lon"][:],
            io_unperturbed["lat"][:],
            io_unperturbed["tcwv"][:, 0, frame].std(axis=0),
            transform=ccrs.PlateCarree(),
            cmap="RdPu",
            levels=std_levels_perturbed,
            norm=LogNorm(vmin=1e-1, vmax=std_levels_perturbed[-1]),
        )
        ax1.coastlines()
        ax1.gridlines()

        im2 = ax2.contourf(
            io_perturbed["lon"][:],
            io_perturbed["lat"][:],
            io_perturbed["tcwv"][:, 0, frame].mean(axis=0),
            transform=ccrs.PlateCarree(),
            cmap="Blues",
            levels=levels_perturbed,
        )
        ax2.coastlines()
        ax2.gridlines()

        im3 = ax3.contourf(
            io_perturbed["lon"][:],
            io_perturbed["lat"][:],
            io_perturbed["tcwv"][:, 0, frame].std(axis=0),
            transform=ccrs.PlateCarree(),
            cmap="RdPu",
            levels=std_levels_perturbed,
            norm=LogNorm(vmin=1e-1, vmax=std_levels_perturbed[-1]),
        )
        ax3.coastlines()
        ax3.gridlines()

        for i in range(16):
            ax0.contour(
                io_unperturbed["lon"][:],
                io_unperturbed["lat"][:],
                io_unperturbed["z500"][i, 0, frame] / 100.0,
                transform=ccrs.PlateCarree(),
                levels=np.arange(485, 580, 15),
                colors="black",
                linestyle="dashed",
            )

            ax2.contour(
                io_perturbed["lon"][:],
                io_perturbed["lat"][:],
                io_perturbed["z500"][i, 0, frame] / 100.0,
                transform=ccrs.PlateCarree(),
                levels=np.arange(485, 580, 15),
                colors="black",
                linestyle="dashed",
            )
        plt.suptitle(
            f'Forecast Starting on {forecast_date} - Lead Time - {io_perturbed["lead_time"][frame]}'
        )

        if frame == 0:
            ax0.set_title("Unperturbed Ensemble Mean - tcwv + z500 countors")
            ax1.set_title("Unperturbed Ensemble Std - tcwv")
            ax2.set_title("Perturbed Ensemble Mean - tcwv + z500 contours")
            ax2.set_title("Perturbed Ensemble Std - tcwv")

            plt.colorbar(
                im0, ax=ax0, shrink=0.75, pad=0.04, label="kg m^-2", format="%2.1f"
            )
            plt.colorbar(
                im1, ax=ax1, shrink=0.75, pad=0.04, label="kg m^-2", format="%1.2e"
            )
            plt.colorbar(
                im2, ax=ax2, shrink=0.75, pad=0.04, label="kg m^-2", format="%2.1f"
            )
            plt.colorbar(
                im3, ax=ax3, shrink=0.75, pad=0.04, label="kg m^-2", format="%1.2e"
            )


    # Uncomment this for animation
    # update(0)
    # ani = animation.FuncAnimation(
    # fig=fig, func=update, frames=range(1, nsteps), cache_frame_data=False
    # )
    # ani.save(f"outputs/model_perturbation_{forecast_date}.gif", dpi=300)

    # Here we plot a handful of images
    for lt in [0, 10, 20, 30, 40]:
        update(lt)
        plt.savefig(
            f"outputs/model_perturbation_{forecast_date}_leadtime_{lt}.png",
            dpi=300,
            bbox_inches="tight",
        )



.. image-sg:: /examples/images/sphx_glr_05_model_perturbation_hook_001.png
   :alt: Forecast Starting on 2024-01-30 - Lead Time - 240 hours
   :srcset: /examples/images/sphx_glr_05_model_perturbation_hook_001.png, /examples/images/sphx_glr_05_model_perturbation_hook_001_2_00x.png 2.00x
   :class: sphx-glr-single-img


.. rst-class:: sphx-glr-script-out

 .. code-block:: none

    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1642: UserWarning: Log scale: values of z <= 0 have been masked
      result = super().contourf(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1642: UserWarning: Log scale: values of z <= 0 have been masked
      result = super().contourf(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /home/jenkins/agent/workspace/Studio-Build-Docs/repo/examples/05_model_perturbation_hook.py:374: UserWarning: Attempt to set non-positive ylim on a log-scaled axis will be ignored.
      plt.colorbar(
    /home/jenkins/agent/workspace/Studio-Build-Docs/repo/examples/05_model_perturbation_hook.py:380: UserWarning: Attempt to set non-positive ylim on a log-scaled axis will be ignored.
      plt.colorbar(
    /usr/local/lib/python3.10/dist-packages/cartopy/io/__init__.py:241: DownloadWarning: Downloading: https://naturalearth.s3.amazonaws.com/50m_physical/ne_50m_coastline.zip
      warnings.warn(f'Downloading: {url}', DownloadWarning)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)
    /usr/local/lib/python3.10/dist-packages/cartopy/mpl/geoaxes.py:1601: UserWarning: The following kwargs were not used by contour: 'linestyle'
      result = super().contour(*args, **kwargs)





.. rst-class:: sphx-glr-timing

   **Total running time of the script:** (0 minutes 44.218 seconds)


.. _sphx_glr_download_examples_05_model_perturbation_hook.py:

.. only:: html

  .. container:: sphx-glr-footer sphx-glr-footer-example

    .. container:: sphx-glr-download sphx-glr-download-jupyter

      :download:`Download Jupyter notebook: 05_model_perturbation_hook.ipynb <05_model_perturbation_hook.ipynb>`

    .. container:: sphx-glr-download sphx-glr-download-python

      :download:`Download Python source code: 05_model_perturbation_hook.py <05_model_perturbation_hook.py>`


.. only:: html

 .. rst-class:: sphx-glr-signature

    `Gallery generated by Sphinx-Gallery <https://sphinx-gallery.github.io>`_
